packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx

```diff
@@ -0,0 +1,213 @@
+ * limitations under the License.
+ */
+
+import {
+  Dialog,
+  PanelDropZone,
+  PanelFormBooleanField,
+  PanelForm,
+  PanelDivider,
+  Modal,
+  ModalBody,
+  ModalFooter,
+  ModalHeader,
+  PanelFormListItems,
+} from '@finos/legend-art';
+import {
+  isSuperType,
+  PrimitiveInstanceValue,
+  PRIMITIVE_TYPE,
+  type ValueSpecification,
+  type VariableExpression,
+} from '@finos/legend-graph';
+import { guaranteeNonNullable } from '@finos/legend-shared';
+import { observer } from 'mobx-react-lite';
+import { useCallback } from 'react';
+import { useDrop } from 'react-dnd';
+
+import {
+  type QueryBuilderParameterDragSource,
+  QUERY_BUILDER_PARAMETER_DND_TYPE,
+} from '../../stores/QueryBuilderParametersState.js';
+import type { QueryBuilderState } from '../../stores/QueryBuilderState.js';
+import type { QueryBuilderWatermarkState } from '../../stores/watermark/QueryBuilderWatermarkState.js';
+import { VariableExpressionViewer } from '../QueryBuilderParametersPanel.js';
+import { BasicValueSpecificationEditor } from '../shared/BasicValueSpecificationEditor.js';
+
+const WatermarkValueEditor = observer(
+  (props: {
+    watermarkState: QueryBuilderWatermarkState;
+    queryBuilderState: QueryBuilderState;
+  }) => {
+    const { watermarkState } = props;
+    const variable = watermarkState.value as VariableExpression;
+    const graph = watermarkState.queryBuilderState.graphManagerState.graph;
+    const parameterType = watermarkState.value?.genericType?.value.rawType;
+
+    const handleDrop = useCallback(
+      (item: QueryBuilderParameterDragSource): void => {
+        watermarkState.setValue(item.variable.parameter);
+      },
+      [watermarkState],
+    );
+
+    const [{ isParameterValueDragOver }, dropTargetConnector] = useDrop<
+      QueryBuilderParameterDragSource,
+      void,
+      { isParameterValueDragOver: boolean }
+    >(
+      () => ({
+        accept: [QUERY_BUILDER_PARAMETER_DND_TYPE],
+        drop: (item, monitor): void => {
+          const itemType = item.variable.parameter.genericType?.value.rawType;
+          if (
+            !monitor.didDrop() &&
+            // Doing a type check, which only allows dragging and dropping parameters of the same type or of child types
+            itemType &&
+            parameterType &&
+            (isSuperType(parameterType, itemType) ||
+              parameterType.name === itemType.name)
+          ) {
+            handleDrop(item);
+          } // prevent drop event propagation to accomondate for nested DnD
+        },
+        collect: (monitor) => ({
+          isParameterValueDragOver: monitor.isOver({
+            shallow: true,
+          }),
+        }),
+      }),
+      [handleDrop],
+    );
+
+    return (
+      <>
+        {watermarkState.value && (
+          <div key={variable.name} className="panel__content__form__section">
+            <div className="panel__content__form__section__header__label">
```

Comment:
This div should not be necessary. The value spec editor should handle rendering value spec of type variable expression.